### PR TITLE
Pyroscope: Add limit number of series returned by the data source

### LINKED ISSUE #93791: Pyroscope: Add ability to limit number of series returned by the datasource
**Why is this needed**:

In some cases the number of returned series may be quite big and it would be good if there was a way to control it. For example Explore Profiles app shows only up to 10 series in the breakdown but all series are queried and sent over the network.

**What would you like to be added**:

A way to limit number of returned series. It could be only at the query level to allow apps to control it but it can also be added to the Query Editor for Explore users.

**Who is this feature for?**

Mainly Explore Profiles app

**Technical details**

This can be implemented in Pyroscope SelectSeries (https://github.com/grafana/pyroscope/issues/3593) or in plugin's backend depending on what makes more sense.

### PR DESCRIPTION
**Pull request by grafakus**
**What is this feature?**

This PR modifies Explore when a Pyroscope data source is selected:
- a new input has been added to limit the number of series to return (when grouping by a specific label)
- a new `limit` parameter is passed to the data source backend, which in turn, forwards it to the new version of the Pyroscope API

image

**Why do we need this feature?**

Generally speaking, to improve the performance when grouping by a specific label. Specifically, to improve the performance of the Explore Profiles application in the "Labels" view (see the [related issue](https://github.com/grafana/explore-profiles/issues/192))

**Who is this feature for?**

All users.

**Which issue(s) does this PR fix?**:

- Supersedes https://github.com/grafana/grafana/pull/94851
- Fixes https://github.com/grafana/grafana/issues/93791
- Blocks https://github.com/grafana/explore-profiles/pull/219

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Comment by ifrost** [comment]
Works great 👍 

There's a small issue with default value:

| Step | Screen |
|--------|--------|
| Initial load: Input shows "20" as default but nothing is passed | Screenshot 2024-11-05 at 11 03 25 |
| Changing to any value correctly changes the query 👍  | Screenshot 2024-11-05 at 11 04 22 |
| Removing the value in the input: shows "20" in the placeholder but passes "0" in query | Screenshot 2024-11-05 at 11 05 09 |

No value or "0" in the query seems to apply no limit at all so maybe we can just change the placeholder?

**Comment by ifrost** [comment]
One more thought, probably something for the future if it becomes a problem for on-prem users. If old backend is used (without limiting) the input is still shwon in the UI. It will have no effect on the query and it makes it hard to figure out why it's not working correctly.

**Comment by ifrost** [comment]
Should we also add some docs to https://github.com/grafana/grafana/blob/e8f23bb2cf666c4b380fc6f9f4e94c0ba5b1b326/docs/sources/datasources/pyroscope/query-profile-data.md#L7? cc @knylander-grafana

**Comment by ifrost** [comment]
**suggestion (non-blocking)**: might be worth adding it to collapsed info as well:

Screenshot 2024-11-05 at 11 16 24

**Review by ivanahuckova — COMMENTED** [review]
Could the changes to parca backend be removed, as they don't seem relevant to this PR.

**Comment by grafakus** [comment]
> Could the changes to parca backend be removed, as they don't seem relevant to this PR.

We actually need them to prevent the error [encountered in the 1st version of this PR](https://github.com/grafana/grafana/pull/94851#discussion_r1804480397). @bryanhuhta can provide more details about it.

**Comment by grafakus** [comment]
> One more thought, probably something for the future if it becomes a problem for on-prem users. If old backend is used (without limiting) the input is still shwon in the UI. It will have no effect on the query and it makes it hard to figure out why it's not working correctly.

I raised the topic with the Pyroscope team. Given our users base, it's not a problem for now but something we want to keep in mind in the near future as our users base grow.

**Comment by grafakus** [comment]
> Should we also add some docs to
> 
> https://github.com/grafana/grafana/blob/e8f23bb2cf666c4b380fc6f9f4e94c0ba5b1b326/docs/sources/datasources/pyroscope/query-profile-data.md#L7
> 
> ? cc @knylander-grafana

My humble attempt: https://github.com/grafana/grafana/pull/95780/commits/79173e2eb02b0c800665c61e9afa12607d1f7d9e

File: docs/sources/datasources/pyroscope/query-profile-data.md [hunk 1]
```diff
@@ -135,6 +135,8 @@ Using **Group by**, you can group metric data by a specified label.
 Without any **Group by** label, metric data aggregates over all the labels into single time series.
 You can use multiple labels to group by. Group by only effects the metric data and doesn't change the profile data results.
 
+In conjunction with **Group by**, you can set a positive number in the **Limit** input to limit the maximum number of time series returned by the data source. Note that the series returned are always ordered by descending value (total aggregated data over the time period).
```
Comment (knylander-grafana) [inline docs/sources/datasources/pyroscope/query-profile-data.md:138]:
```suggestion
In conjunction with **Group by**, you can set a positive number in the **Limit** input to limit the maximum number of time series returned by the data source. The series returned are always ordered by descending value for the total aggregated data over the time period.
```

**Comment by knylander-grafana** [comment]
> Should we also add some docs to
> 
> https://github.com/grafana/grafana/blob/e8f23bb2cf666c4b380fc6f9f4e94c0ba5b1b326/docs/sources/datasources/pyroscope/query-profile-data.md#L7
> 
> ? cc @knylander-grafana

That looks good! Thanks for thinking about linking to docs.

**Comment by bryanhuhta** [comment]
> Could the changes to parca backend be removed, as they don't seem relevant to this PR.

@ivanahuckova The changes required to the Parca backend are subtle, but necessary. I'm happy to entertain alternative solutions. Perhaps that means breaking these changes into two PRs: one to upgrade the Parca backend and one to add the new feature to the Pyroscope backend?

The TL;DR is that the Pyroscope API clients published after (and including) version `v0.4.0` are incompatible with projects that use `github.com/bufbuild/connect-go` (which is what the Parca data source backend used). By switching over to `connectrpc.com/connect`, we resolve that incompatibility.

The longer version is: Pyroscope publishes a Go package ([`github.com/grafana/pyroscope/api`](https://pkg.go.dev/github.com/grafana/pyroscope/api)) which contains all the gRPC models and clients. Critically, the clients are `connect` clients to enable using protobufs over HTTP.

For this PR to work, an existing Pyroscope API endpoint had to be modified. This work was done in https://github.com/grafana/pyroscope/pull/3602. The change to the API definitions https://github.com/grafana/pyroscope/pull/3602 introduced was made available in the Go package in the [`api/v1.0.0` tag](https://github.com/grafana/pyroscope/releases/tag/api%2Fv1.0.0).

In order to take advantage of these new gRPC model definitions, we needed to upgrade the Go package from its current version [`v0.3.0`](https://github.com/grafana/pyroscope/releases/tag/api%2Fv0.3.0) to [`v1.0.0`](https://github.com/grafana/pyroscope/releases/tag/api%2Fv1.0.0).

However, a side effect of this upgrade means we skip over [`v0.4.0`](https://github.com/grafana/pyroscope/releases/tag/api%2Fv0.4.0), which changed the library that generated the `connect` clients from `github.com/bufbuild/connect-go` to `connectrpc.com/connect` as the original bufbuild library got donated to the CNCF under connectrpc.

If you try build a project using both `github.com/bufbuild/connect-go` and `connectrpc.com/connect`, you'll get an error like:

```
panic: proto: file "connectext/grpc/status/v1/status.proto" is already registered
    previously from: "connectrpc.com/connect/internal/gen/connectext/grpc/status/v1"
    currently from:  "github.com/bufbuild/connect-go/internal/gen/connectext/grpc/status/v1"
See https://protobuf.dev/reference/go/faq#namespace-conflict
```

This occurs because both `github.com/bufbuild/connect-go` and `connectrpc.com/connect` have differently generated protobuf stubs of the exact same protobuf definitions. This can be suppressed by building with this flag:

```
-X "google.golang.org/protobuf/reflect/protoregistry.conflictPolicy=warn"
```

but we eventually rejected that approach in https://github.com/grafana/grafana/pull/94851 because it felt like opening ourselves up to bugs that would be tricky to predict, debug, and understand.

The approach we landed on was using the [`connectrpc/go`](https://buf.build/connectrpc/go?version=v1.17.0) bufbuild plugin instead of the deprecated [`bufbuild/connect-go`](https://buf.build/bufbuild/connect-go?version=v1.10.0) plugin.

Thus, we come full circle and have a PR here which largely modifies the Pyroscope data source backend, but in doing so has a knock-on effect on the Parca data source backend. That said, the changes to the Parca data source backend shouldn't have any functional impact and should be equivalent to upgrading a dependency.

**Comment by ivanahuckova** [comment]
Thank you @bryanhuhta for detailed explanation. Really appreciate it and makes sense. So you can disregard my comments. 🙂

File: public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/QueryOptions.tsx [hunk 2]
```diff
@@ -47,6 +47,9 @@ export function QueryOptions({ query, onQueryChange, app, labels }: Props) {
   if (query.groupBy?.length) {
     collapsedInfo.push(`Group by: ${query.groupBy.join(', ')}`);
   }
+  if (query.limit) {
+    collapsedInfo.push(`Limi: ${query.limit}`);
```
Comment (bryanhuhta) [inline public/app/plugins/datasource/grafana-pyroscope-datasource/QueryEditor/QueryOptions.tsx:51]:
```suggestion
    collapsedInfo.push(`Limit: ${query.limit}`);
```

**Review by bryanhuhta — APPROVED** [review]
LGTM

**Review by ifrost — APPROVED** [review]
LGGGGGGGTM 👍

**Pull request merged**